### ISSUE TITLE
Scan or map_fn function to speed up custom RNN cell runtime

### ISSUE THREAD

**Issue by windweller** [pytorch/pytorch#4275]
Hi,

I think this is related also to this issue: Zoneout LSTM (https://github.com/pytorch/pytorch/issues/561)

There are many instances where a custom RNN cell needs to be used:
Zoneout LSTM, Variational Dropout, etc.

Also to quote https://github.com/pytorch/pytorch/pull/1234 Smerity's comment.

Python is notoriously slow in its for-loop structure, and if a simple function that just executes Lua/C for-loop, even without CUDA optimization, should be faster than Python solutions.

A function similar to Theano's scan or Tensorflow's map_fn would be very helpful!

**Comment by apaszke** [comment]
`tensor.apply_(lambda x: x + 1)`. I wouldn't expect it to be too fast though.

**Comment by windweller** [comment]
Thanks!
